File: Technology Fundamentals/14.Exercises - Array Advanced/train.js
function solve(input){
    let  train = input.shift();
    train =  train.split(' ');
    train =  train.map(el => Number(el));

    let maxCapacity = Number(input.shift());
    
    for (let index = 0; index < input.length; index++) {
        let command = input[index].split(' ')[0];
        let passengers = undefined;
        if(command == 'Add'){
            passengers = Number(input[index].split(' ')[1]);
            train.push(passengers);
        }else{
            passengers = Number(input[index].split(' ')[0]);
            for (let index = 0; index < train.length; index++) {
                const element = train[index];
                let countPassengers = element + passengers;
                
                if (countPassengers <= maxCapacity){
                    train[index] = countPassengers;
                    break;
                }
            }
        }
    }
    console.log(train.join(' '));
}
solve(['32 54 21 12 4 0 23', '75', 
'Add 10', 'Add 0', '30', '10', '75']);

File: Technology Fundamentals/14.Exercises - Array Advanced/Csharp/01.Train/Train.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace _01.Train
{
    public class Train
    {
        public static void Main()
        {
            List<int> wagons = Console.ReadLine()
                .Split(' ')
                .Select(int.Parse)
                .ToList();

            int maxCapacity = int.Parse(Console.ReadLine());

            string command = Console.ReadLine();

            while (command != "end")
            {
                string[] items = command.Split();
                int passengers = 0;
                if (items[0] == "Add")
                {
                    passengers = int.Parse(items[1]);
                    wagons.Add(passengers);
                }
                else
                {
                    passengers = int.Parse(items[0]);
                    AddPassengersInWagon(wagons, maxCapacity, passengers);
                }
                
                command = Console.ReadLine();
            }

            PrintStateWagon(wagons);
        }

        private static void PrintStateWagon(List<int> wagons)
        {
            Console.WriteLine(string.Join(" ", wagons));
        }

        private static void AddPassengersInWagon(List<int> wagons, int maxCapacity, int passengers)
        {
            for (int i = 0; i < wagons.Count; i++)
            {
                int wagon = wagons[i];
                if (IsHaveFreePlace(maxCapacity, passengers, wagon))
                {
                    wagons[i] += passengers;
                    break;
                }
            }
        }

        private static bool IsHaveFreePlace(int maxCapacity, int passengers, int wagon)
        {
            return wagon + passengers <= maxCapacity;
        }
    }
}
